src/Domain/CoordinateMaps/FocallyLiftedMapHelpers.cpp

```diff
@@ -0,0 +1,215 @@
+
+namespace domain::CoordinateMaps::FocallyLiftedMapHelpers {
+
+template <typename T>
+tt::remove_cvref_wrap_t<T> scale_factor(
+    const std::array<T, 3>& src_point, const std::array<double, 3>& proj_center,
+    const std::array<double, 3>& sphere_center, double radius,
+    const bool src_is_between_proj_and_target) noexcept {
+  using return_type = tt::remove_cvref_wrap_t<T>;
+  // quadratic equation is
+  // a x^2 + b x + c = 0
+  const return_type a = square(src_point[0] - proj_center[0]) +
+                        square(src_point[1] - proj_center[1]) +
+                        square(src_point[2] - proj_center[2]);
+  const return_type b =
+      2.0 *
+      ((src_point[0] - proj_center[0]) * (proj_center[0] - sphere_center[0]) +
+       (src_point[1] - proj_center[1]) * (proj_center[1] - sphere_center[1]) +
+       (src_point[2] - proj_center[2]) * (proj_center[2] - sphere_center[2]));
+  const double c = square(sphere_center[0] - proj_center[0]) +
+                   square(sphere_center[1] - proj_center[1]) +
+                   square(sphere_center[2] - proj_center[2]) - square(radius);
+  if (src_is_between_proj_and_target) {
+    // Here we assume that src_point is between proj_center and
+    // target_point.  There are three cases: 1) src and proj are both
+    // inside the sphere, 2) src and proj are both outside the sphere,
+    // and 3) proj is outside the sphere and src is inside the sphere.
+    // To cover all 3 cases, we choose the smallest root that is
+    // greater than or equal to unity. This means for case 2) we are
+    // choosing the point closest to src.
+    return smallest_root_greater_than_value_within_roundoff(
+        a, b, make_with_value<return_type>(a, c), 1.0);
+  }
+  // Here we assume that target_point is between proj_center and
+  // src_point. There are three cases: 1) proj is inside the sphere
+  // and src is outside the sphere 2) src is inside the sphere and proj
+  // is outside the sphere, and 3) the sphere is between src and proj.
+  // To cover all 3 cases, we choose the largest root that is less than
+  // or equal to unity, and we require that this root is positive.
+  // This means that for 3) we are choosing the point closest to src.
+  return largest_root_between_values_within_roundoff(
+      a, b, make_with_value<return_type>(a, c), 0.0, 1.0);
+}
+
+boost::optional<double> try_scale_factor(
+    const std::array<double, 3>& src_point,
+    const std::array<double, 3>& proj_center,
+    const std::array<double, 3>& sphere_center, double radius,
+    const bool pick_larger_root,
+    const bool pick_root_greater_than_one) noexcept {
+  // We solve the quadratic for (scale_factor-1) instead of scale_factor to
+  // avoid roundoff problems when scale_factor is very nearly equal to unity.
+  // Note that scale_factor==1 will occur when src_point is on the sphere, which
+  // happens when inverse-mapping the boundaries.
+
+  // quadratic equation is
+  // a x^2 + b x + c = 0
```

Comment:
Please add a derivation of these terms to the help text? I checked that they are correct, but it's subtle, because if you forget that you're solving for scale_factor - 1, a is unchanged and b and c are almost unchanged